### PR TITLE
Allow custom contexts!

### PR DESCRIPTION
**Pull request by Vixeliz**
This will allow custom context to be made for even more customization of how ggez works! Of course a default context will still be provided. Minimum changes to the user api due to this change has occured.

### CODE REVIEW

**Comment by Vixeliz** [comment]
I forgot to change all the examples over I will do that in a new commit here

File: src/event.rs [hunk 1]
```diff
@@ -95,24 +101,25 @@ pub enum ErrorOrigin {
 ///
 /// For the error type simply choose the default [`GameError`](../error/enum.GameError.html),
 /// or something more generic, if your situation requires it.
-pub trait EventHandler<E = GameError>
+pub trait EventHandler<C, E = GameError>
 where
     E: std::fmt::Debug,
+    C: std::any::Any + HasMut<ContextFields> + HasMut<input::mouse::MouseContext>,
```
Comment (nobbele) [inline src/event.rs:107]: why the Any bound?

File: src/event.rs [hunk 2]
```diff
@@ -95,24 +101,25 @@ pub enum ErrorOrigin {
 ///
 /// For the error type simply choose the default [`GameError`](../error/enum.GameError.html),
 /// or something more generic, if your situation requires it.
-pub trait EventHandler<E = GameError>
+pub trait EventHandler<C, E = GameError>
```
Comment (nobbele) [inline src/event.rs:104]: Default this to the standard context type

File: src/context.rs [hunk 3]
```diff
@@ -344,17 +399,44 @@ impl ContextBuilder {
...
+        from_conf: F,
+    ) -> GameResult<(C, winit::event_loop::EventLoop<()>)>
+    where
+        F: Fn(String, conf::Conf, Filesystem) -> GameResult<(C, winit::event_loop::EventLoop<()>)>,
+        C: std::any::Any
+            + HasMut<ContextFields>
+            + HasMut<timer::TimeContext>
+            + HasMut<input::keyboard::KeyboardContext>
+            + HasMut<input::mouse::MouseContext>
+            + HasMut<GamepadContext>,
```
Comment (nobbele) [inline src/context.rs:418]: Should probably make
```rust
macro_rules! trait_alias {
    ($name:ident = $base1:ident + $($base2:ident +)+) => {
        trait $name: $base1 $(+ $base2)+ { }
        impl<T: $base1 $(+ $base2)+> $name for T { }
    };
}

// The last + is actually needed
trait_alias!(HasMutCommonContexts, HasMut<ContextFields> + /* ... */ + HasMut<timer::TimeContext> +)
```

**Review by nobbele — CHANGES_REQUESTED** [review]

File: src/event.rs (hunk 1 above)
Comment (Vixeliz) [reply src/event.rs:107]: Ah that was before I added the other bounds kind of slipped my mind. I'll fix that up

**Review by nobbele — APPROVED** [review]
Nice

**Pull request merged**